src/mods/validate.cpp

```diff
@@ -0,0 +1,214 @@
+    bool dpad_down =
+        pad::button_down(mkb::PAD_BUTTON_DOWN) || pad::button_down(mkb::PAD_BUTTON_LEFT) ||
+        pad::button_down(mkb::PAD_BUTTON_RIGHT) || pad::button_down(mkb::PAD_BUTTON_UP);
+    if (pref::get(pref::BoolPref::DpadControls) && dpad_down) s_used_mods = true;
+
+    // Opening the mod menu is disallowed
+    if (menu_impl::is_visible()) s_used_mods = true;
+
+    // Physics must be default or custom (with all default values)
+    if (physics::using_custom_physics()) s_used_mods = true;
+
+    // Invalid bool prefs are enabled
+    for (u8 i = 0; i < LEN(INVALID_BOOL_PREFS); i++) {
+        if (pref::get(INVALID_BOOL_PREFS[i]) != pref::get_default(INVALID_BOOL_PREFS[i])) {
+            s_used_mods = true;
+        }
+    }
+    // Invalid u8 prefs are enabled
+    for (u8 i = 0; i < LEN(INVALID_U8_PREFS); i++) {
+        if (pref::get(INVALID_U8_PREFS[i]) != pref::get_default(INVALID_U8_PREFS[i])) {
+            s_used_mods = true;
+        }
+    }
+}
+
+// Copy-paste of line_intersects from ghidra with slight changes to calculate s_framesave
+static bool line_intersects(const Vec& lineStart, const Vec& lineEnd, const mkb::Rect& rect) {
+    Vec end;
+    Vec start;
+    float half_height;
+    float half_width;
+
+    start.x = lineStart.x;
+    start.y = lineStart.y;
+    start.z = lineStart.z;
+    end.x = lineEnd.x;
+    end.y = lineEnd.y;
+    end.z = lineEnd.z;
+    mkb::mtxa_from_translate(&const_cast<Vec&>(rect.pos));
+    mkb::mtxa_rotate_z((rect.rot).z);
+    mkb::mtxa_rotate_y((rect.rot).y);
+    mkb::mtxa_rotate_x((rect.rot).x);
+    mkb::mtxa_rigid_inv_tf_point(&start, &start);
+    mkb::mtxa_rigid_inv_tf_point(&end, &end);
+    if (((end.z < 0.0) && (start.z < 0.0)) || ((0.0 < end.z && (0.0 < start.z)))) {
+        return false;
+    } else {
+        half_width = start.z - end.z;
+        if (1.192093e-07 < half_width) {
+            end.x = end.x - (start.x - end.x) * (end.z / half_width);
+            end.y = end.y - (start.y - end.y) * (end.z / half_width);
+        }
+        half_width = rect.width * 0.5;
+        half_height = rect.height * 0.5;
+        if ((end.x < -half_width) || (half_width < end.x)) {
+            return false;
+        } else if ((end.y < -half_height) || (half_height < end.y)) {
+            return false;
+        } else {
+            // update framesave if first goal entered
+            if (mkb::sub_mode != mkb::SMD_GAME_GOAL_INIT &&
+                mkb::sub_mode != mkb::SMD_GAME_GOAL_MAIN &&
+                mkb::sub_mode != mkb::SMD_GAME_GOAL_REPLAY_INIT &&
+                mkb::sub_mode != mkb::SMD_GAME_GOAL_REPLAY_MAIN) {
+                s_framesave = static_cast<u32>((start.z / (start.z - end.z)) * 100);
+            }
+            return true;
+        }
+    }
+}
+
+// Copy paste from ghidra of did_ball_enter_goal with slight changes to calculate s_framesave
+void find_framesave(mkb::Ball* ball, int* out_stage_goal_idx, int* out_itemgroup_id,
```

Comment:
Framesave calculation seems pretty orthogonal to IL validation and maybe ought to go in a different file eventually, but I'll let it slide for this PR - it makes sense to factor it out of timer.cpp at least